src/platform/qt/library/LibraryController.cpp

```diff
@@ -0,0 +1,162 @@
+/* Copyright (c) 2014-2017 waddlesplash
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+#include "LibraryController.h"
+
+#include "../GBAApp.h"
+#include "LibraryGrid.h"
+#include "LibraryTree.h"
+
+namespace QGBA {
+
+CLibraryEntry::CLibraryEntry(mLibraryEntry* entry)
+    : entry(entry)
+    , m_fullpath(QString("%1/%2").arg(entry->base, entry->filename))
+{
+}
+CLibraryEntry::~CLibraryEntry() {
+}
+
+void AbstractGameList::addEntries(QList<LibraryEntry> items) {
+	for (LibraryEntry o : items) {
+		addEntry(o);
+	}
+}
+void AbstractGameList::removeEntries(QList<LibraryEntry> items) {
+	for (LibraryEntry o : items) {
+		addEntry(o);
+	}
+}
+
+LibraryController::LibraryController(QWidget* parent_widget, const QString& path, ConfigController* config)
+	: QStackedWidget(parent_widget)
+	, m_config(config)
+{
+	mLibraryListingInit(&m_listing, 0);
+
+	if (!path.isNull()) {
+		m_library = mLibraryLoad(path.toUtf8().constData());
+	} else {
+		m_library = mLibraryCreateEmpty();
+	}
+
+	mLibraryAttachGameDB(m_library, GBAApp::app()->gameDB());
+
+	m_libraryTree = new LibraryTree(this);
+	addWidget(m_libraryTree);
+	connect(m_libraryTree, &LibraryTree::StartGame, this, &LibraryController::startGame);
+
+	m_libraryGrid = new LibraryGrid(this);
+	addWidget(m_libraryGrid);
+	connect(m_libraryGrid, &LibraryGrid::StartGame, this, &LibraryController::startGame);
+
+	refresh();
+	selectLastBootedGame(); // TODO: doesn't belong here
+
+	setViewStyle(LibraryStyle::STYLE_TREE);
+}
+
+LibraryController::~LibraryController() {
+	mLibraryListingDeinit(&m_listing);
+	if (m_library) {
+		mLibraryDestroy(m_library);
+	}
+}
+
+void LibraryController::setViewStyle(LibraryStyle newStyle) {
+	m_currentStyle = newStyle;
+
+	AbstractGameList* newCurrentWidget = nullptr;
+	if (newStyle == LibraryStyle::STYLE_LIST || newStyle == LibraryStyle::STYLE_TREE) {
+		newCurrentWidget = m_libraryTree;
+	} else {
+		newCurrentWidget = m_libraryGrid;
+	}
+	newCurrentWidget->selectEntry(selectedEntry());
+	newCurrentWidget->setViewStyle(newStyle);
+	setCurrentWidget(dynamic_cast<QWidget*>(newCurrentWidget));
```

Comment:
I'm not too familiar with diamond inheritance, but inheriting QWidget on AbstractGameList would avoid all of these dynamic casts. That being said, composition is better than inheritance, and it's probably entirely possible to do this without MI. Hell, a bunch of the functions are copy/paste implementations over slightly different types, so templates may even work.